core/chains/evm/forwarders/forwarder_manager.go

```diff
@@ -0,0 +1,291 @@
+		latestBlock:  0,
+	}
+	return &fwdMgr
+}
+
+// Start starts Forwarder Manager.
+func (f *FwdMgr) Start() error {
+	return f.StartOnce("EVMForwarderManager", func() error {
+		f.logger.Debug("Initializing EVM forwarder manager")
+
+		fwdrs, err := f.ORM.FindForwardersByChain(utils.Big(*f.evmClient.ChainID()))
+		if err != nil {
+			return errors.Wrapf(err, "Failed to retrieve forwarders for chain %d", f.evmClient.ChainID())
+		}
+		f.ctx, f.cancel = context.WithCancel(context.Background())
+		if len(fwdrs) != 0 {
+			f.initForwardersCache(f.ctx, fwdrs)
+			f.subscribeForwardersLogs(fwdrs)
+		}
+
+		f.authRcvr, err = authorized_receiver.NewAuthorizedReceiver(common.Address{}, f.evmClient)
+		if err != nil {
+			return errors.Wrap(err, "Failed to init AuthorizedReceiver")
+		}
+
+		f.offchainAgg, err = offchain_aggregator_wrapper.NewOffchainAggregator(common.Address{}, f.evmClient)
+		if err != nil {
+			return errors.Wrap(err, "Failed to init OffchainAggregator")
+		}
+
+		f.wg.Add(1)
+		go f.runLoop()
+		return nil
+	})
+}
+
+// TODO(samhassan): this should be aware of job type to decide how to fetch senders list.
+func (f *FwdMgr) MaybeForwardTransaction(from common.Address, to common.Address, encodedPayload []byte) (fwdAddr common.Address, fwdPayload []byte, err error) {
+
+	senders, err := f.getContractSenders(to)
+	if err != nil {
+		return to, encodedPayload, errors.Wrap(err, "Skipping forwarding transaction")
+	}
+
+	// Gets current forwarders that are in `to` senders
+	fwdrs, err := f.ORM.FindForwardersInListByChain(utils.Big(*f.evmClient.ChainID()), senders)
+	if err != nil {
+		return to, encodedPayload, errors.Wrap(err, "Skipping forwarding transaction")
+	}
+
+	for _, fwdr := range fwdrs {
+		eoas, err := f.getContractSenders(fwdr.Address)
+		if err != nil {
+			f.logger.Warnf("Failed to get forwarder senders %s", err)
+			continue
+		}
+		for _, eoa := range eoas {
+			if eoa != from {
+				continue
+			}
+			forwardedPayload, err := f.getForwardedPayload(to, encodedPayload)
+			if err != nil {
+				f.logger.AssumptionViolation("Forwarder encoding failed, this should never happen")
+				continue
+			}
+			return fwdr.Address, forwardedPayload, nil
+		}
+	}
+
+	return to, encodedPayload, errors.New("Skipping forwarding transaction")
+}
+
+func (f *FwdMgr) getForwardedPayload(dest common.Address, origPayload []byte) ([]byte, error) {
+	callArgs, err := forwardABI.Inputs.Pack(dest, origPayload)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to pack forwarder payload")
+	}
+
+	dataBytes := append(forwardABI.ID, callArgs...)
+	return dataBytes, nil
+}
+
+func (f *FwdMgr) getContractSenders(addr common.Address) ([]common.Address, error) {
+	if senders, ok := f.getCachedSenders(addr); ok {
+		return senders, nil
+	}
+	senders, err := f.getAuthorizedSenders(addr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to call getAuthorizedSenders on %s", addr)
+	}
+	f.setCachedSenders(addr, senders)
```

Comment:
Still seems like this can race

Reply:
Can you elaborate on this one?

Reply:
I don't see a data race - the internals look syncronized. But if it is possible for two concurrent calls to `getContractSenders` to have the same `addr`, then they could race to update this cache, and the technically 'worse/older' result could win. However, this is no worse than the calls occuring just a momement in time sooner, when they both would fetch the older results, so doesn't seem necessarily problematic.

This could be related to a large problem though: do we depend on receiving events to keep this cache from going stale in general? We may want to periodically refresh after a TTL or something :shrug:

Reply:
So the way I see it, we couldn't gone either the TTL route or the events route, we don't expect these caches to be invalidated frequently at all. This event listening logic is just a protection mechanism to be backward compatible with contracts with no forwarder setup in place. I've been discussing with steve that going forward, creation, management of forwarders (and hooking them up to jobs) should be done within the node on behalf of the Node Operator (more precisely the multiple EOAs on a node) this way we can drop listening to onchain events and the forwarder logic can potentially become much simpler than this imho

Reply:
That sounds good long term.

But what happens now if we miss events, and this becomes stale? Will the node operator have to notice some error(s) and intervene with a reboot or something? If so, could we instead automate that detection and refresh/recover? If it cannot be auto-repaired, then a TTL backup would at least change the worst case from being unbounded to some fixed upper bound.

Reply:
Maybe a good mitigation is to invalidate the cache on reverted forwarded txs? i.e when a tx thats going through a forwarder fail, instruct fwdr to invalidate cache. This is what I was thinking of moving forward.